2016/presentations/async_pod_maska/text.md

```diff
@@ -0,0 +1,476 @@
+modelu kooperacyjnym, jest uruchamianie schedulera za każdym razem, gdy program
+prosi system operacyjny o jakiś rodzaj operacji wejścia-wyjścia, np. o
+odczytanie stanu klawiatury albo wyświetlenie tekstu na ekranie. W tym celu
+dostarczymy specjalne warianty tych funkcji, w naszym przykładzie funkcji
+`print`:
+
+    def smart_print(*args, **kwargs):
+        print(*args, **kwargs)
+        # tutaj chcemy uruchomić scheduler
+
+    def process(pid):
+        for i in itertools.count():
+            smart_print(pid, i)
+
+Pozostaje zatem pytanie, jak uruchomić scheduler. Jeśli po prostu zawołamy
+go jako funkcję, to po jej zakończeniu wrócimy najpierw do `smart_print` a
+potem do tego samego procesu. Z kolei jeśli w `smart_print` użyjemy
+`yield`, to przerwiemy tylko funkcję `smart_print`, znów wracając do `process`.
+Potrzebujemy mechanizmu, który pozwoli poprosić inny generator, aby wykonał
+`yield` za nas i zatrzymać działanie głównego generatora. Wykonanie programu
+powinno przebiegać jak na poniższym diagramie:
+
+![yield from](yield_from.png)
+
+Jest to wykonalne, ale kod który realizuje taką operację jest niezwykle
+skomplikowany, gdyż musi brać pod uwagę, że `os_print` może rzucić wyjątkiem,
+zawierać kilka `yield`ów (np. w pętli) i tak dalej. Ostatecznie jest to ponad
+30 linii kodu [2] które musielibyśmy umieszczać w programie przy *każdej* funkcji
+wejścia-wyjścia. Byłoby to skrajnie niepraktyczne.
+
+Python 3.4 rozwiązuje ten problem dostarczając nowe słowo kluczowe `yield
+from`, które wykonuje całą sekwencję za nas.
+
+    def os_print(*args, **kwargs):
+        print(*args, **kwargs)
+        yield
+
+    def this_is_a_process(pid):
+        for i in itertools.count():
+            yield from os_print(pid, i)
+
+Na marginesie, warto zauważyć, że `yield from` bywa użyteczne też w innych
+kontekstach, na przykład gdy chcemy napisać rekurencyjny generator:
+
+    def walk(dir):
+        for i in os.listdir(dir):
+            path = os.path.join(dir, i)
+            yield path
+            if os.path.isdir(path):
+                yield from walk(path)
+
+Programy wyświetlające napisy na ekranie nie są jednak zbyt ekscytujące, a o
+programowaniu asynchronicznym rozmawia się zazwyczaj w konktekście obsługi
+sieci. Żeby zobaczyć związek między jednym a drugiem, musimy odłożyć na chwilę
+problem przełączania procesów, a zająć się obsługą wejścia-wyjścia na
+poziomie systemu operacyjnego.
+
+## Deskryptory plików i gniazda sieciowe
+
+Praktycznie wszystkie systemy operacyjne udostępniają operacje wejścia-wyjścia
+za pośrednictwem *deskryptorów plików*. Deskryptor jest zazwyczaj po
+prostu liczbą, którą dostajemy od systemu w momencie otwarcia pliku, a
+przekazujemy go jako argument do funkcji, które z tego pliku czytają bądź do
+niego piszą.
+
+W ten sposób deskryptor jednoznacznie identyfikuje plik, który chcemy obsłużyć.
+Nie interesuje nas za to jego wartość liczbowa deskryptora, używamy go raczej
```

Comment:
s/Nie interesuje nas za to jego wartość liczbowa deskryptora/Nie interesuje nas wartość liczbowa deskryptora